Require dependencies to exist during sorting phase

paf31 opened the pull request:
Fixes #1921

I did some tidying up again as I went through here. Sorry for all the commit noise.

The main changes are:

- Adding the `ModuleNotFound` error, and throwing it during module sorting.
- Simplifying the module sorting logic, since we only need to look at imports now, I think.

It turns out the problem was that if a dependency were deleted, it was never getting passed into `Make` in the first place, so the "should rebuild" test was always returning `False`, since `Make` thought the other module had no dependencies.

Since `sortModules` is the first thing to happen, it makes sense to me that we should check for missing imports there.

paf31 commented:
It seems this messes up fast rebuild somehow, which I can verify using `pscid`. Thankfully the test suite caught it 😄 

I'll look more tomorrow.

kRITZCREEK commented on src/Language/PureScript/ModuleDependencies.hs:
```diff
@@ -1,99 +1,65 @@
-    in (m, getModuleName m, nub (concatMap (usedModules ams) ds))
+    toGraphNode :: [ModuleName] -> Module -> m (Module, ModuleName, [ModuleName])
+    toGraphNode mns m@(Module _ _ mn ds _) = do
+      let deps = nub (concatMap usedModules ds)
```
This could be very expensive for lists with 30 modulenames that start with: "Slamdata.Editor". ordNub is just better.

kRITZCREEK commented on src/Language/PureScript/ModuleDependencies.hs:
```diff
@@ -1,99 +1,65 @@
-  where
+-- | Calculate a list of used modules based on explicit imports and qualified names.
+usedModules :: Declaration -> [ModuleName]
+usedModules d = nub (f d) where
```
another `nub` on module names

garyb replied:
We could just use a `Set` here anyway perhaps? I think there are a few places we use lists where sets would better express the intention and probably perform better also.

paf31 commented:
@kRITZCREEK I've addressed the test failure and verified it using `pscid` locally. I also updated the `Integration.hs` module so that we can see error messages when a test fails.

I agree that we probably shouldn't be using `nub`, but I'd like to leave that for later, for two reasons:

- It's not high in the profile right now anyway
- I'm planning to work on performance issues in general soon.

kRITZCREEK commented:
> It's not high in the profile right now anyway

Equality on `Name`'s is in all the profiles though? Isn't that an indication of `nub` creating performance problems?

rvion commented:
There is quite a few nub as of now: https://github.com/purescript/purescript/search?p=3&q=nub&utf8=%E2%9C%93

paf31 commented:
Updated. Better?

hdgarrood commented:
Did you use stylish-haskell on the files you modified during this PR? Would it make sense to run that (or whichever tool it was) on all the Haskell source files in this repo and add a note to CONTRIBUTING.md that any changes to source files should run the same tool on any changed source files too?

paf31 commented:
No, I've been formatting things by hand. I haven't found a template which will generate things how I want to see them, and I haven't had time to create one, but maybe I should ..
